### PR TITLE
Make dispatcher registrations of SymInt functions backwards compatible

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84579
* __->__ #84557

Previously, when we SymInt-ify a schema, this is a BC-breaking change
for all people who registered functions for that function; they
must accept c10::SymInt where they previously accepted int64_t.
This is not great.

With this change, I accept old type registrations transparently.  The
idea is in several parts:

- At the registration site, at compile time I have no idea whether or not
  if the function being registered has a SymInt schema or not.  So I
  must defer the exact compatibility check.  What I do instead is
  check if the function pointer registered to me has SymInt in the
  argument or not.  If it does, I assume it is new-style and ensure
  it is also registered to a special sym_ slot on KernelFunction.
  If not, it only goes in the conventional slot.

- At the dispatcher site, I know at compile time whether or not this
  is a SymInt function.  If it is, I check for a sym_ slot on the
  KernelFunction, and preferentially use that.  If no such slot
  exists, I then fall back to the regular slot... but I convert
  all SymInt arguments to int64_t arguments (doing assertions that
  no true symbolic integer was passed.)  I can skip this test entirely
  if the function doesn't have any SymInts in it; in that case I know
  that only the original slot could have been registered. Fortunately,
  both branches of the short circuit typecheck, so I didn't have to
  use SFINAE or if-constexpr to make it work; just a plain if statement
  that I expect the compiler to optimize away.

- Schema validation is now modestly more complicated. There are two parts. First, function schema validation proceeds by checking if the signature in question has any SymInt-like types in it or not. If it does, we do function schema validation against the real types; if it doesn't, we do validation against the fake types (but only for symint; MemoryFormat is always MemoryFormat). Second, cpp signature validation also keeps track of a "symint" cpp signature and a "non-symint" cpp signature. We only compare symint with symint, and non-symint with non-symint. I did not implement checking a conflict between a symint and non-symint cpp signature, though in principle you could try converting the SymInt types to non-SymInt types and doing the comparison that way.

To show it is working, I remove a bunch of c10::asIntArrayRefSlow shims, as the dispatcher is able to insert them automatically now.

I didn't update the Metal registrations (though they can get similar treatment) as OSS CI coverage is insufficient for this case.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D39280965](https://our.internmc.facebook.com/intern/diff/D39280965)

### CODE REVIEW

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39280965).

File: aten/src/ATen/core/boxing/KernelFunction_impl.h [hunk 1]
```diff
@@ -43,16 +50,46 @@ inline Return callUnboxedKernelFunction(void* unboxed_kernel_func, OperatorKerne
...
     // want callers to explicitly specify the Args.
 
-    if (C10_LIKELY(unboxed_kernel_func_ != nullptr)) {
-      auto *functor = boxed_kernel_func_.getFunctor();
-      return callUnboxedKernelFunction<Return, Args...>(
-          unboxed_kernel_func_, functor, dispatchKeySet, std::forward<Args>(args)...);
+    // This should get inlined by compiler
+    if (guts::disjunction<has_symint<Args>...>::value) {
+      if (sym_unboxed_kernel_func_ != nullptr) {
+        auto *functor = boxed_kernel_func_.getFunctor();
```
Comment (wconstab) [inline aten/src/ATen/core/boxing/KernelFunction_impl.h:77]: is the getFunctor call so we can get some python state associated with the boxed kernel and then use that same state with the sym_unboxed kernel?  I'm pretty unfamiliar here

Comment (ezyang) [reply aten/src/ATen/core/boxing/KernelFunction_impl.h:83]: No, this is a twisty corner of the boxing implementation. Functor refers to a rarely used piece of functionality in KernelFunction which allows operators to be defined as arbitrary Functor classes that can have extra state in them (much like how a lambda that closes over variables has extra state). Closure-less lambdas and regular function pointers are functors in a degenerate sense, they have no state. Everything goes through functor, so I updated the functor path only. It doesn't really matter, just mentally substitute functor with function and the code is still meaningful.

File: aten/src/ATen/core/dispatch/OperatorEntry.cpp [hunk 2]
```diff
@@ -34,12 +35,11 @@ OperatorEntry::OperatorEntry(OperatorName&& operator_name)
...
 namespace {
-  void checkSchema(const OperatorName& name, const FunctionSchema& from_def, const std::string& from_def_debug, const FunctionSchema& inferred, const std::string& inferred_debug) {
+  void checkSchema(const OperatorName& name, const FunctionSchema& from_def_, const std::string& from_def_debug, const KernelFunction& kernel, const FunctionSchema& inferred_, const std::string& inferred_debug) {
     // TODO: figure out if we can just directly save real schema at def time
-    c10::optional<std::string> schema_difference = findSchemaDifferences(
-      from_def.cloneWithRealTypes(),
-      inferred.cloneWithRealTypes()
-    );
+    FunctionSchema from_def = from_def_.cloneWithRealTypes(kernel.isValidSymUnboxed());
+    FunctionSchema inferred = inferred_.cloneWithRealTypes();
```
Comment (wconstab) [inline aten/src/ATen/core/dispatch/OperatorEntry.cpp:41]: so, if the C++ function decl had SymInt, the yaml must also be updated

if the yaml had SymInt, boxed or Unboxed-nonsymint kernels are allowed to persist for BC.  (and presumably somewhere there is codegen to convert the symint to int64)

ok this part looks right to me

File: aten/src/ATen/core/boxing/KernelFunction_impl.h [hunk 3]
```diff
@@ -43,16 +50,52 @@ inline Return callUnboxedKernelFunction(void* unboxed_kernel_func, OperatorKerne
...
+        return callUnboxedKernelFunction<Return, Args...>(
+            sym_unboxed_kernel_func_, functor, dispatchKeySet, std::forward<Args>(args)...);
+      }
+
+      if (unboxed_kernel_func_ != nullptr) {
+        auto *functor = boxed_kernel_func_.getFunctor();
+        return callUnboxedKernelFunction<Return, typename remove_symint<Args>::type...>(
```
Comment (wconstab) [inline aten/src/ATen/core/boxing/KernelFunction_impl.h:90]: ahh ok, missed this first time, 

so if we are given symint args and we don't have a symint kernel, we use the remove_symint macros to convert to int and call the non-symint kernel.

**Review by wconstab — APPROVED** [review]
This looks good to me.

File: aten/src/ATen/core/dispatch/OperatorEntry.cpp (hunk 2 above)
Comment (ezyang) [reply aten/src/ATen/core/dispatch/OperatorEntry.cpp:41]: Yes.

File: aten/src/ATen/core/boxing/KernelFunction_impl.h (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/core/boxing/KernelFunction_impl.h:90]: Yep!

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39280965).

**Pull request closed without merging**